src/assets/Debugging/debugging.md

```diff
@@ -0,0 +1,378 @@
+---
+title: Advanced Docker Container Debugging - A Comprehensive Guide
+authorName: Prajwol Amatya
+authorAvatar: https://1.gravatar.com/avatar/de64e53c0e2cb393dd0d14ffdd53058ee9c607b35e366dd392425bd1b95a034c?size=256
+authorLink: https://github.com/prajwolamatya
+createdAt: May 07, 2025
+tags: debugging, docker
+banner: https://blog.jankaritech.com/src/assets/Debugging/images/Debugging.png
+---
+
+Docker has revolutionized modern software development by enabling lightweight, portable, and scalable containerized applications. However, as deployments grow in complexity, so do the challenges in debugging and optimizing containers. In this blog, we'll explore powerful debugging techniques using a sample project with Node.js, Nginx, and Redis. You can get the sample project [here](https://github.com/prajwolamatya/debug-docker).
+
+Our setup consists of:
+- A Node.js application (port 3000)
+- Nginx as a reverse proxy (port 80)
+- Redis for caching (port 6379)
+
+## 1. Container Inspection
+### Viewing Running Containers
+First, let's check our running containers:
+
+```bash
+docker compose ps
+```
+
+**Expected Output:**
+```console
+Name                       Command                         State    Ports
+debug-docker_nginx_1       /docker-entrypoint.sh ngin ...  Up       0.0.0.0:80->80/tcp
+debug-docker_nodejs-app_1  docker-entrypoint.sh node ...   Up       0.0.0.0:3000->3000/tcp
+debug-docker_redis_1       docker-entrypoint.sh redis ...  Up       0.0.0.0:6379->6379/tcp
+```
+
+- **State:** Shows if container is running/stopped
+- **Ports:** Reveals port mappings (host:container)
+
+### Inspecting Container Details
+For deeper inspection of a specific container:
+
+```bash
+docker inspect debug-docker_nodejs-app_1
+```
+
+This returns a JSON with all container details including:
+- Network settings
+- Mounts
+- Environment variables
+- IP addresses
+
+**Filter specific information:**
+
+```bash
+docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' debug-docker_nodejs-app_1
+```
+
+Let's breakdown the command:
+- `-f` or `--format` activates Go template formatting
+- `{{range .NetworkSettings.Networks}}` iterates through networks
+- `{{.IPAddress}}` extracts the IP for each network
+- `{{end}}` closes the loop
+
+**Example Output:**
+
+(*The container's internal IP in Docker's network*)
+```console
+172.19.0.3
+```
+
+Docker uses **Go Templates** for the `--format` filtering in `docker inspect`. You can get more details on how to construct the filter in Go's [text/template](https://pkg.go.dev/text/template) package.
+Also, you can learn more on formatting output in [Format command and log output](https://docs.docker.com/engine/cli/formatting/) section.
+
+## 2. Log Analysis
+### Viewing Container Logs
+
+```bash
+docker compose logs
+```
+
+This will return the logs of all the running continers. You can also get the container specific logs using the following command.
```

Comment:
The log command will also show the logs of containers that are part of the docker compose setup, but exited.